membraneframework/membrane-element-aac
# AAC Decoder

## Description
Adds base foundation elements for AAC Decoder.

Heavily inspired by: https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/libfdk-aacdec.c

## Changes
* Added basic native interface for AAC operations that uses https://github.com/mstorsjo/fdk-aac library
* Implemented placeholder Elixir methods that call Native interface

## Test plan
* Create a test pipeline that uses `.aac` file as input source
* Connect `AAC.Decoder` inside that pipeline
* Ensure all frames are correctly decoded

## TODO
Note: The implementation is still a work in progress. ~I'm currently facing `0x2007` error from FDK-AAC library (`AAC_DEC_UNSUPPORTED_CHANNELCONFIG`?) or a segfault.~ However, the interface definitions should not change that much.
- [x] Fix native interface errors and ensure the frames are parsed correctly
- [ ] Handle more configuration options
- [x] Return explicit errors
- [x] ~Implement buffers queue~ (No need, FDK-AAC has its own buffer)
- [x] ~Handle EndOfStream event and cleanup native decoder state accordingly~ (No need, we're reading all frames in `handle_process`)
- [x] ~Add 2112 samples delay? (according to https://developer.apple.com/library/archive/technotes/tn2258/_index.html)~ (Handled internally by FDK-AAC)
- [x] Use Membrane's logging methods
- [x] Add docstrings
- [x] Handle `valid != 0` by throwing an explicit error? (Will handle that in a separate PR as it is not obvious what to do in this case)
